extensions/ql-vscode/src/query-server/query-server-client.ts

```diff
@@ -240,8 +240,32 @@ export class QueryServerClient extends DisposableObject {
     this.nextCallback = 0;
     this.nextProgress = 0;
     this.progressCallbacks = {};
-    child.on("close", () => {
-      this.restartQueryServerOnFailure();
+
+    // 'exit' may or may not fire after 'error' event, so we want to guard against restarting the
+    // query server twice if both events fire.
+    let wasExitOrErrorHandled = false;
+    child.on("error", (err: Error) => {
+      if (!wasExitOrErrorHandled) {
+        void this.logger.log(`Query server terminated with error: ${err}.`);
+        this.restartQueryServerOnFailure();
+        wasExitOrErrorHandled = true;
+      }
+    });
+    child.on("exit", (code: number, signal: string) => {
```

Comment:
Just making sure: can `close` ever fire without `exit` also firing?

Reply:
Yes, but only in the `error` case:

> The 'close' event will always emit after ['exit'](https://nodejs.org/api/child_process.html#event-exit) was already emitted, or ['error'](https://nodejs.org/api/child_process.html#event-error) if the child failed to spawn.